misc/diagram-util/src/main/java/org/ballerinalang/diagramutil/SyntaxTreeMapGenerator.java

```diff
@@ -200,20 +197,57 @@ protected JsonElement transformSyntaxNode(Node node) {
             }
 
             nodeJson.add("typeData", symbolJson);
-            if (node.kind() == SyntaxKind.FUNCTION_BODY_BLOCK
-                    && (this.visibleEpsForEachBlock.size() > 0 || this.visibleEpsForModule.size() > 0)
-                    && nodeJson.get("typeData") != null) {
-                JsonArray eps = new JsonArray();
-                this.visibleEpsForEachBlock.forEach(eps::add);
-                this.visibleEpsForModule.forEach(eps::add);
-                nodeJson.add("VisibleEndpoints", eps);
-                this.visibleEpsForEachBlock = new ArrayList<>();
+
+            if ((node.kind() == SyntaxKind.BLOCK_STATEMENT || node.kind() == SyntaxKind.FUNCTION_BODY_BLOCK)
+                    && (this.visibleEpsForEachBlock.size() > 0 || this.visibleEpsForModule.size() > 0)) {
+
+                JsonArray blockEndpoints = new JsonArray();
+                // Add module level endpoints
+                this.visibleEpsForModule.forEach(blockEndpoints::add);
+
+                if (this.visibleEpsForEachBlock.size() > 0) {
+                    for (JsonObject endpoint : this.visibleEpsForEachBlock) {
+                        int epStartLine = endpoint.get("position").getAsJsonObject().get("startLine").getAsInt();
+                        int epEndLine = endpoint.get("position").getAsJsonObject().get("endLine").getAsInt();
+
+                        Optional<Node> parentFunctionBlock = getParentFunctionBlock(node);
+                        if (parentFunctionBlock.isPresent()
+                                && epStartLine >= parentFunctionBlock.get().lineRange().startLine().line()
+                                && epEndLine < node.lineRange().startLine().line()
+                                && !endpoint.has("innerBlock")) {
+                            blockEndpoints.add(endpoint);
+                        }
+
+                        if (epStartLine >= node.lineRange().startLine().line()
+                                && epEndLine < node.lineRange().endLine().line()
+                                && !endpoint.has("innerBlock")) {
+                            blockEndpoints.add(endpoint);
+                            // Add key to filter endpoints only visible to a block
+                            endpoint.addProperty("innerBlock", true);
+                        }
+                    }
+                }
+                nodeJson.add("VisibleEndpoints", blockEndpoints);
```

Comment:
won't this logic fails to rightfully detect what shown to the block level ? coz I didn't see a line where we remove endpoints that already been classified as belong inside the block ... So if we have a code like below
```
function name() {
  <endpoint1>
   if true {
      <endpoint2>
  }
}
```
as visible endpoint classification happens at the end won't `endpoint2` will be added to the visible list of function block as it is still in the visible endpoint list and it is between function block start and end?

Reply:
This block will add all block-level endpoints and mark them as an `inner block` endpoint. https://github.com/ballerina-platform/ballerina-lang/blob/257bef32df4f35bf2c1c48eacb40fbdc574d5ff6/misc/diagram-util/src/main/java/org/ballerinalang/diagramutil/SyntaxTreeMapGenerator.java#L221-L227

And all other endpoints above the selected block and inside the function block will detect by the below condition. `getParentFunctionBlock` function will recursively get the function body node to detect the function body scope. https://github.com/ballerina-platform/ballerina-lang/blob/257bef32df4f35bf2c1c48eacb40fbdc574d5ff6/misc/diagram-util/src/main/java/org/ballerinalang/diagramutil/SyntaxTreeMapGenerator.java#L213-L219 
I have validated inner level endpoints using the `interlock` field.

Reply:
ack